pkg/reconciler/reconciler.go

```diff
@@ -0,0 +1,75 @@
+package reconciler
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type Reconciler interface {
+	Reconcile(ctx context.Context, c client.Client, scheme *runtime.Scheme) error
+}
+
+type UpdateReconciler struct {
+	controller metav1.Object
+	resource   client.Object
+}
+
+func (r UpdateReconciler) Reconcile(ctx context.Context, c client.Client, scheme *runtime.Scheme) error {
+	if r.controller.GetNamespace() == r.resource.GetNamespace() {
+		if err := controllerutil.SetControllerReference(r.controller, r.resource, scheme); err != nil {
+			return err
+		}
+	}
+
+	if err := c.Patch(ctx, r.resource, client.Apply, client.ForceOwnership, client.FieldOwner(fmt.Sprintf("%s/%s", r.controller.GetNamespace(), r.controller.GetName()))); err != nil {
+		return err
+	}
+	return nil
+}
+
+func NewUpdateReconciler(r client.Object, c metav1.Object) UpdateReconciler {
+	return UpdateReconciler{
+		controller: c,
+		resource:   r,
+	}
+}
+
+type DeleteReconciler struct {
+	resource client.Object
+}
+
+func (r DeleteReconciler) Reconcile(ctx context.Context, c client.Client, scheme *runtime.Scheme) error {
+	if err := c.Delete(ctx, r.resource); client.IgnoreNotFound(err) != nil {
+		return err
+	}
+	return nil
+}
+
+func NewDeleteReconciler(r client.Object) DeleteReconciler {
+	return DeleteReconciler{resource: r}
+}
+
+type OptionalResourceReconciler struct {
+	r Reconciler
+}
+
+func (r OptionalResourceReconciler) Reconcile(ctx context.Context, c client.Client, scheme *runtime.Scheme) error {
+	return r.r.Reconcile(ctx, c, scheme)
+}
+
+func NewOptionalResourceReconciler(r client.Object, c metav1.Object, cond bool) OptionalResourceReconciler {
```

Comment:
Since the function already decides which reconciler to use, can't we  remove the need for an Optional reconciler and instead return the concrete reconciler itself ? e.g.

```
func NewDeleterOrUpdater(cond bool, obj client.Object, ctrl metav1.Object) Reconciler {
   if cond {
     return Deleter{obj}
   }
   return Updater{obj, ctrl}
}
```
The caller would call 

```
reconciler.NewDeleterOrUpdater(ms.Spec.Replicas >= 2, newAlertmanager(), ctrl)
```

Reply:
I agree with @sthaha, it feels simpler.